server/listener.go

```diff
@@ -137,11 +133,22 @@ func (l *Listener) Close() error {
 			return err
 		}
 	}
-	err = l.eg.Wait()
-	close(l.conns)
-	return err
+	if l.shutdown != nil {
```

Comment:
it's better to have this check guarded by the mutex. otherwise... one go routine could check this, see the `l.shutdown` is not nil, but there could be another go routine holding the mutex in `CloseChannels` calling close on `l.shutdown`. Then the first go routine wakes back up, calls `CloseChannels` and panics because it calls close on a nil reference.

Reply:
I would just call `CloseChannels` here and let it do all the touching of the sensitive/shared bits. Alternatively, you could grab the mutex at the top of the `Close` method instead and not need to have the `CloseChannels` function.